### DISCUSSION TITLE
Trying to add a button to the toolbar

### DISCUSSION THREAD

**Discussion by marty-a11y** [storybookjs/storybook#28690] [Help] [answered]
### Summary

I'm trying to add a simply button to the toolbar. It needs not to be a default dropdown button - I already know how to successfully add those. I've been told by several sources that I need to create a small addon and register that. So I made this:

`/.storybook/addons/mock-switch.tsx`:
```tsx
import React from 'react';
import { addons, types } from '@storybook/manager-api';
import { IconButton } from '@storybook/components';
import { CheckIcon } from '@storybook/icons';

const ADDON_ID = 'mock-switch';
const TOOL_ID = `${ADDON_ID}/toolbar`;

console.log('Registering addon');

addons.register(ADDON_ID, (api) => {
  console.log('Registering tool');

  addons.add(TOOL_ID, {
    type: types.PANEL,
    title: 'Minimal Addon',
    render: ({ active }) => (
       <IconButton
        key={TOOL_ID}
        title="Mock switch"
      >
        <CheckIcon />
      </IconButton>
    ),
  });
});
```

`/.storybook/main.ts`:
```typescript
import type { StorybookConfig } from "@storybook/react-vite";

const config: StorybookConfig = {
  stories: ["../src/**/*.mdx", "../src/**/*.stories.@(js|jsx|mjs|ts|tsx)"],
  addons: [
    "@storybook/addon-onboarding",
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@chromatic-com/storybook",
    "@storybook/addon-interactions",
    "./addons/mock-switch",
  ],
  framework: {
    name: "@storybook/react-vite",
    options: {},
  },
};
export default config;
```

The above `main.ts` has only got the addon added to an otherwise new, clean, bog standard react-based empty storybook project.

To clarify, I just did `npx sb init` in a totally empty directory, and when with all the default options, making sure to pick React/Vite as the framework. Then I added the above code to try and add a addon with as little amount of code as possible to illustrate the problem.

### What's the problem

Glad you asked :)

The log `Registering addon` appears, but `Registering tool` does not. In other words, it appears as though the `addons.register` is being called perfectly well (and I can prove it by calling it twice - it will complain about a duplicate registration) but its callback is never called, which gives my addon the chnce to register its tool exactly never (for now at least).

Here is the entire log output in the CLI:

```
>npm run storybook

> storybook-minimal-addon@0.0.0 storybook
> storybook dev -p 6006

storybook v8.2.5

Registering addon
info => Starting manager..
info => Starting preview..
info Using tsconfig paths for react-docgen
╭────────────────────────────────────────────────────╮
│                                                    │
│   Storybook 8.2.5 for react-vite started           │
│   253 ms for manager and 642 ms for preview        │
│                                                    │
│    Local:            http://localhost:6006/        │
│    On your network:  http://10.110.68.129:6006/    │
│                                                    │
╰────────────────────────────────────────────────────╯
Sourcemap for "/virtual:/@storybook/builder-vite/setup-addons.js" points to missing source files
Sourcemap for "/virtual:/@storybook/builder-vite/vite-app.js" points to missing source files
```

In the browser console, nothing worth noting appears.

I'm sure I'm doing something wrong, but I can't figure out for the life of me what it is.

### Additional information

Sorry I'm unfamiliar with stackblitz. I hope I done it right.

### Create a reproduction

https://stackblitz.com/edit/github-y8dhdo?file=.storybook%2Faddons%2Fmock-switch.tsx,.storybook%2Fmain.ts,tsconfig.json&preset=node

**Comment by Marklb** [comment, marked as answer]
The docs go into more details, describing the expected way of writing an addon: https://storybook.js.org/docs/addons/writing-addons

I modified your Stackblitz, with two ways of registering that I don't think are specifically documented and I don't know enough about addons to confirm that they are stable solutions. Also, I also changed your addon from a Panel to a Tool, since that is what it sounds like you wanted.
https://stackblitz.com/edit/github-y8dhdo-owsdlu

Your Stackblitz just got hung while loading, for me, with how you were adding the file to your addons config. I think it did actually work that way, in a previous version, but I don't remember how it did, without it knowing to use manager or preview.

I don't think is is mentioned in the docs, since it is focused on addons that are built and published as a package. The docs are more specific, but the build output of an addon is basically a `preview.js` or `manager.js` or both. The filename specifies where it will be imported. So, for your code, I created a folder, with a `manager.ts` file and specified that folder in `.storybook/main.ts`. I can't guarantee it is the right thing to do, since I rarely add my own addon code and don't know it very well, but in my projects I typically just use `.storybook/manager.ts` or `.storybook/preview.ts` for small api code.

**Reply by marty-a11y** [reply]
Hey that's pretty cool, it works! Thanks!

Yes it needed to be a Tool rather than a Panel. That was probably a remnant of me fiddling with it for a while until I decided to open a discussion 😀

In my actual project I did get a WSOD but that was quickly solved by adding a `import React from 'react';` in the tool. I'm not sure why this is needed - I thought this wasn't neccesary anymore in React 18. But it works, and that's what counts.

So, might I be so bold to claim the documentation probably needs a bit of freshening up in terms of addon authoring? It seems the documentation is *nearly* correct, doesn't it.
